internal/engine/ast.go

```diff
@@ -253,14 +253,39 @@ func mkConstStringExpr(s string) *exprpb.Expr {
 	}
 }
 
-func mkListExpr(elems []*exprpb.Expr) *exprpb.Expr {
-	return &exprpb.Expr{
-		ExprKind: &exprpb.Expr_ListExpr{
-			ListExpr: &exprpb.Expr_CreateList{
-				Elements: elems,
-			},
-		},
+func structKeys(x *exprpb.Expr_CreateStruct) *exprpb.Expr_CreateList {
+	type element struct {
+		*exprpb.Expr
+		SortKey string
+	}
+
+	elems := make([]element, len(x.Entries))
+	for i, entry := range x.Entries {
+		var expr *exprpb.Expr
+
+		switch e := entry.KeyKind.(type) {
+		case *exprpb.Expr_CreateStruct_Entry_MapKey:
+			expr = e.MapKey
+		case *exprpb.Expr_CreateStruct_Entry_FieldKey:
+			expr = mkConstStringExpr(e.FieldKey)
+		}
+
+		elems[i] = element{
+			Expr:    expr,
+			SortKey: (&exprpb.Expr{ExprKind: expr.ExprKind}).String(), // sort by expression, ignoring AST node id
+		}
+	}
+
+	sort.Slice(elems, func(i, j int) bool {
+		return elems[i].SortKey < elems[j].SortKey
+	})
+
+	exprs := make([]*exprpb.Expr, len(elems))
+	for i, elem := range elems {
+		exprs[i] = elem.Expr
 	}
+
+	return &exprpb.Expr_CreateList{Elements: exprs}
```

Comment:
Might make sense to move the construction of the list expression in here too? Both places we use `structKeys` we immediately wrap it like 

```go
&exprpb.Expr{
	ExprKind: &exprpb.Expr_ListExpr{
		ListExpr: structKeys(x),
	},
}
```

Reply:
I wanted to highlight the returning type of the structKeys function. I've refactored this snippet a bit.